ambari-server/src/main/java/org/apache/ambari/server/stack/UpdateActiveRepoVersionOnStartup.java

```diff
@@ -88,10 +88,14 @@ public void process() throws AmbariException {
 
           StackId stackId = repositoryVersion.getStackId();
           StackInfo stack = stackManager.getStack(stackId.getStackName(), stackId.getStackVersion());
-
-          if (updateRepoVersion(stack, repositoryVersion)) {
-            repositoryVersionDao.merge(repositoryVersion);
-          }
+        
+          if(stack !=null) {
+            if (updateRepoVersion(stack, repositoryVersion)) {
+              repositoryVersionDao.merge(repositoryVersion);
+            }
+          } else {
+            LOG.error(String.format("Check if Stack %s  version %s is present in file system",  stackId.getStackName(), stackId.getStackVersion()));
```

Comment:
I think an exception should be thrown here.  Isn't is really bad if the stack is `null`?

```
throw new AmbariException("Stack {} version {} was not found in file system",  stackId.getStackName(), stackId.getStackVersion())
```

Reply:
@rlevas  thanks for reviewing it. In subsequent calls there is an stack exception if there is no nullpointerexception here. here is the sample log.
-------------------
2019-03-18 16:49:10,611 ERROR [main] UpdateActiveRepoVersionOnStartup:97 - Check if Stack HDP  version 3.1 is present in file system
2019-03-18 16:49:10,615  INFO [main] CertificateManager:75 - Initialization of root certificate
2019-03-18 16:49:10,616  INFO [main] CertificateManager:77 - Certificate exists:true
2019-03-18 16:49:10,616  INFO [main] KerberosChecker:128 - Skipping Ambari Server Kerberos credentials check.
2019-03-18 16:49:10,960  INFO [main] ContextLoader:304 - Root WebApplicationContext: initialization started
2019-03-18 16:49:10,961  INFO [main] AnnotationConfigWebApplicationContext:583 - Refreshing Root WebApplicationContext: startup date [Mon Mar 18 16:49:10 UTC 2019]; parent: org.springframework.context.support.ClassPathXmlApplicationContext@1b5f960a
2019-03-18 16:49:10,983  INFO [main] AnnotationConfigWebApplicationContext:208 - Registering annotated classes: [class org.apache.ambari.server.configuration.spring.ApiSecurityConfig]
2019-03-18 16:49:11,052  INFO [main] AmbariManagementControllerImpl:427 - Initializing the AmbariManagementControllerImpl
2019-03-18 16:49:11,128  INFO [main] SingleFileWatch:78 - Starting SingleFileWatcher:ambari.properties
2019-03-18 16:49:11,131  INFO [main] AmbariServer:786 - Jetty is configuring ambari-client-thread with 2 reserved acceptors/selectors and a total pool size of 25 for 32 processors.
2019-03-18 16:49:11,149  INFO [main] AmbariServer:786 - Jetty is configuring qtp-ambari-agent with 4 reserved acceptors/selectors and a total pool size of 25 for 32 processors.
2019-03-18 16:49:11,228  INFO [main] ClustersImpl:281 - Initializing cluster and host data.
2019-03-18 16:49:11,260 ERROR [main] ClusterImpl:389 - Service config versioning disabled due to exception:
org.apache.ambari.server.ParentObjectNotFoundException: Parent Stack Version resource doesn't exist.  Stack data, Stack HDP 3.1 is not found in Ambari metainfo.  Stack data, Stack HDP 3.1 is not found in Ambari metainfo
        at org.apache.ambari.server.api.services.AmbariMetaInfo.getServices(AmbariMetaInfo.java:512)
        at org.apache.ambari.server.state.cluster.ClusterImpl.collectServiceConfigTypesMapping(ClusterImpl.java:387)
        at org.apache.ambari.server.state.cluster.ClusterImpl.loadServiceConfigTypes(ClusterImpl.java:370)
        at org.apache.ambari.server.state.cluster.ClusterImpl.loadStackVersion(ClusterImpl.java:2660)
        at org.apache.ambari.server.state.cluster.ClusterImpl.<init>(ClusterImpl.java:348)
        at org.apache.ambari.server.state.cluster.ClusterImpl$$EnhancerByGuice$$d2f9db14.<init>(<generated>)
        at org.apache.ambari.server.state.cluster.ClusterImpl$$EnhancerByGuice$$d2f9db14$$FastClassByGuice$$838904f0.newInstance(<generated>)
        at com.google.inject.internal.ProxyFactory$ProxyConstructor.newInstance(ProxyFactory.java:265)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:111)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:90)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:268)
        at com.google.inject.internal.InjectorImpl$2$1.call(InjectorImpl.java:1019)
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1085)
        at com.google.inject.internal.InjectorImpl$2.get(InjectorImpl.java:1015)
        at com.google.inject.assistedinject.FactoryProvider2.invoke(FactoryProvider2.java:776)
        at com.sun.proxy.$Proxy11.create(Unknown Source)
        at org.apache.ambari.server.state.cluster.ClustersImpl.loadClustersAndHosts(ClustersImpl.java:300)
        at org.apache.ambari.server.state.cluster.ClustersImpl.safelyLoadClustersAndHosts(ClustersImpl.java:266)
        at org.apache.ambari.server.state.cluster.ClustersImpl.getClustersByName(ClustersImpl.java:182)
        at org.apache.ambari.server.state.cluster.ClustersImpl.getClusters(ClustersImpl.java:677)
        at org.apache.ambari.server.checks.DatabaseConsistencyCheckHelper.collectConfigGroupsWithoutServiceName(DatabaseConsistencyCheckHelper.java:1168)
        at org.apache.ambari.server.checks.DatabaseConsistencyCheckHelper.checkConfigGroupsHasServiceName(DatabaseConsistencyCheckHelper.java:1194)
        at org.apache.ambari.server.checks.DatabaseConsistencyCheckHelper.runAllDBChecks(DatabaseConsistencyCheckHelper.java:203)
        at org.apache.ambari.server.controller.AmbariServer.runDatabaseConsistencyCheck(AmbariServer.java:717)
        at org.apache.ambari.server.controller.AmbariServer.run(AmbariServer.java:339)
        at org.apache.ambari.server.controller.AmbariServer.main(AmbariServer.java:1108)
Caused by: org.apache.ambari.server.StackAccessException: Stack data, Stack HDP 3.1 is not found in Ambari metainfo
        at org.apache.ambari.server.api.services.AmbariMetaInfo.getStack(AmbariMetaInfo.java:645)
        at org.apache.ambari.server.api.services.AmbariMetaInfo.getServices(AmbariMetaInfo.java:510)
        ... 25 more
-------------------

Still do you think throwing Ambari exception is optimal?

Reply:
I am not sure.  Maybe @jonathan-hurley, @ncole, or @dlysnichenko can comment.

Reply:
The DB is in a bad state if you have services attached to a cluster with a non-existent stack.  That is the problem that should be solved.

Reply:
I would tend to agree - we shouldn't be skipping this b/c of `null` ... there's an underlying problem causing it. Should we throw an exception here and stop the `ambari-server start` command?

Reply:
@ncole thanks for replying - usually we have seen these kind of issues where customers install/upgrade/remove HDF mpacks - if there is no corresponding mpack/stack they it fails with this exception - so throwing error would give some indication that stack definition is not present in Ambari.

Reply:
In that case, I agree that an exception should be thrown instead.

Reply:
sure - will make the code changes and re-attach it. thanks.

Reply:
@ncole , @rlevas I made the code changes to incorporate ambariexception and and tested again - Can you please review and merge